File: 백준/Silver/2468. 안전 영역/안전 영역.py
import sys

sys.setrecursionlimit(15000)


def dfs(x, y, k):
  for i in range(4):
    nx = x + dx[i]
    ny = y + dy[i]
    if (0 <= nx < n) and (
        0 <= ny < n) and visited[nx][ny] == False and graph[nx][ny] > k:
      visited[nx][ny] = True
      dfs(nx, ny, k)
  return True
  return False


n = int(sys.stdin.readline().rstrip())
graph = []
dx = [0, 0, -1, 1]
dy = [-1, 1, 0, 0]

for _ in range(n):
  graph.append(list(map(int, sys.stdin.readline().rstrip().split())))

max_value = max(map(max, graph))

result = 1
for k in range(1, max_value + 1):
  sub_result = 0
  visited = [[False] * n for _ in range(n)]
  for i in range(n):
    for j in range(n):
      if visited[i][j] == False and graph[i][j] > k:
        visited[i][j] = True
        dfs(i, j, k)
        sub_result += 1
  result = max(sub_result, result)

print(result)

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.io.*;
import java.util.*;

public class Main {
  static int n;
  static int[][] graph;
  static boolean[][] visited;

  static void bfs(int x, int y, int h) {
    Queue<int[]> queue = new LinkedList<>();
    visited[x][y] = true;
    queue.offer(new int[] { x, y });

    int[] dx = { -1, 1, 0, 0 };
    int[] dy = { 0, 0, -1, 1 };
    while (!queue.isEmpty()) {
      int[] curr = queue.poll();
      for (int i = 0; i < 4; i++) {
        int nx = curr[0] + dx[i];
        int ny = curr[1] + dy[i];
        if (nx < 0 || nx >= n || ny < 0 || ny >= n)
          continue;
        if (graph[nx][ny] <= h || visited[nx][ny])
          continue;
        visited[nx][ny] = true;
        queue.offer(new int[] { nx, ny });
      }
    }
  }
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    n = Integer.parseInt(br.readLine());
    graph = new int[n][n];
    int minH = 0;
    int maxH = 0;
    for (int i = 0; i < n; i++) {
      StringTokenizer st = new StringTokenizer(br.readLine());
      for (int j = 0; j < n; j++) {
        graph[i][j] = Integer.parseInt(st.nextToken());
        minH = Math.min(minH, graph[i][j]);
        maxH = Math.max(maxH, graph[i][j]);
      }
    }
  
    int result = 0;
    for (int f = minH; f <= maxH; f++) {
      visited = new boolean[n][n];
      int cnt = 0;
      for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
          if (graph[i][j] > f && !visited[i][j]) {
            bfs(i, j, f);
            cnt++;
          }
        }
      }
      result = Math.max(result, cnt);
    }
    System.out.println(result);
  }
}